### PR TITLE
fix(etcd): handle etcd compaction

### LINKED ISSUE #2680: bug:config_etcd did not handle the watcher be canceled
### Issue description
Current version, the config_etcd watch different prefixes and maintain revisions separately, when some prefix be modified, the others can not perceive revision change. If etcd enable auto compacted,  the watcher with old revision will be canceled because revision be compacted.
hope the watcher can handle the canceled result,  re-pull the latest revision and re-watch

In current version(2.0), canceled response will case panic
2020/11/10 10:23:25 [error] 46#0: *16354 [lua] config_etcd.lua:448: failed to fetch data from etcd: 

> /usr/local/apisix/apisix/core/etcd.lua:115: bad argument #1 to 'ipairs' (table expected, got nil)
stack traceback:
        [C]: in function 'ipairs'
        /usr/local/apisix/apisix/core/etcd.lua:115: in function 'waitdir'
        /usr/local/apisix/apisix/core/config_etcd.lua:255: in function 'sync_data'
        /usr/local/apisix/apisix/core/config_etcd.lua:424: in function </usr/local/apisix/apisix/core/config_etcd.lua:414>
        [C]: in function 'xpcall'
        /usr/local/apisix/apisix/core/config_etcd.lua:414: in function </usr/local/apisix/apisix/core/config_etcd.lua:405>,  etcd key: /apisix/plugin_metadata, context: ngx.timer

### Environment
- apisix version (cmd: apisix version): 2.0
- OS:

### PR DESCRIPTION
**Pull request by spacewander**
Close #2680.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

**Comment by ShiningRush** [comment]
We can make a test case for the situation.

**Comment by spacewander** [comment]
@ShiningRush 
It is too tough to provide a stable test case. I have to run several times manually to reproduce the issue.

**Comment by ShiningRush** [comment]
I think it may not so tough, it should reproduce by follow steps:
- update the some key
- manually compact etcd by etcdctl
- restart etcd to cause apisix reconnect


I haven't actually tried it yet, but I guess it should be able to reproduce stably

**Comment by spacewander** [comment]
@ShiningRush 
Yes, it is how I reproduce it. But in the local machine, the sync is so fast that you don't have enough time to reconnect before sync the data. (This is why I need to try many times manually). It would be worse in the CI environment.

File: apisix/core/etcd.lua [hunk 1]
```diff
@@ -118,6 +118,14 @@ function _M.watch_format(v3res)
     v2res.body = {
         node = {}
     }
+
+    local compact_revision = v3res.result.compact_revision
+    if compact_revision and tonumber(compact_revision) > 0 then
+        -- When the revisions are compacted, there might be compacted changes
```
Comment (membphis) [inline apisix/core/etcd.lua:124]: add a todo comment, missing test cases, will add later

Comment (spacewander) [reply apisix/core/etcd.lua:124]: @membphis 
Solved.

**Review by moonming — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Pull request merged**